flow/layers/layer_unittests.cc

```diff
@@ -0,0 +1,170 @@
+  EXPECT_TRUE(parent.tree_reads_surface());
+}
+
+void TestClipRect(Clip clip_behavior, bool ret) {
+  ClipRectLayer layer = ClipRectLayer(SkRect::MakeWH(5, 5), clip_behavior);
+  layer.Add(ReadbackLayer::Make(true, false));
+  EXPECT_EQ(layer.tree_reads_surface(), ret);
+}
+
+TEST(Layer, ClipRectSaveLayer) {
+  //  TestClipRect(Clip::none, true);     // ClipRectLayer asserts !Clip::none
+  TestClipRect(Clip::hardEdge, true);
+  TestClipRect(Clip::antiAlias, true);
+  TestClipRect(Clip::antiAliasWithSaveLayer, false);
+}
+
+void TestClipRRect(Clip clip_behavior, bool ret) {
+  SkRRect r_rect = SkRRect::MakeRect(SkRect::MakeWH(5, 5));
+  ClipRRectLayer layer = ClipRRectLayer(r_rect, clip_behavior);
+  layer.Add(ReadbackLayer::Make(true, false));
+  EXPECT_EQ(layer.tree_reads_surface(), ret);
+}
+
+TEST(Layer, ClipRRectSaveLayer) {
+  //  TestClipRRect(Clip::none, true);     // ClipRRectLayer asserts !Clip::none
+  TestClipRRect(Clip::hardEdge, true);
+  TestClipRRect(Clip::antiAlias, true);
+  TestClipRRect(Clip::antiAliasWithSaveLayer, false);
+}
+
+void TestClipPath(Clip clip_behavior, bool ret) {
+  SkPath path = SkPath();
+  path.moveTo(0, 0);
+  path.lineTo(5, 0);
+  path.lineTo(0, 5);
+  path.close();
+  ClipPathLayer layer = ClipPathLayer(path, clip_behavior);
+  layer.Add(ReadbackLayer::Make(true, false));
+  EXPECT_EQ(layer.tree_reads_surface(), ret);
+}
+
+TEST(Layer, ClipPathSaveLayer) {
+  //  TestClipPath(Clip::none, true);     // ClipRRectLayer asserts !Clip::none
+  TestClipPath(Clip::hardEdge, true);
+  TestClipPath(Clip::antiAlias, true);
+  TestClipPath(Clip::antiAliasWithSaveLayer, false);
+}
+
+TEST(Layer, ColorFilterSaveLayer) {
+  sk_sp<SkColorFilter> filter = SkColorFilters::LinearToSRGBGamma();
+  ColorFilterLayer layer = ColorFilterLayer(filter);
+  layer.Add(ReadbackLayer::Make(true, false));
+  EXPECT_FALSE(layer.tree_reads_surface());
+}
+
+TEST(Layer, OpacitySaveLayer) {
+  OpacityLayer layer = OpacityLayer(10, SkPoint::Make(0, 0));
+  layer.Add(ReadbackLayer::Make(true, false));
+  EXPECT_FALSE(layer.tree_reads_surface());
+}
+
+void TestPhysicalShapeLayer(Clip clip_behavior, bool ret) {
+  SkPath path = SkPath();
+  path.moveTo(0, 0);
+  path.lineTo(5, 0);
+  path.lineTo(0, 5);
+  path.close();
+  PhysicalShapeLayer layer = PhysicalShapeLayer(
+      SK_ColorRED, SK_ColorBLUE, 1.0f, 100.0f, 10.0f, path, clip_behavior);
+  layer.Add(ReadbackLayer::Make(true, false));
+  EXPECT_EQ(layer.tree_reads_surface(), ret);
+}
+
+TEST(Layer, PhysicalShapeSaveLayer) {
+  TestPhysicalShapeLayer(Clip::none, true);
+  TestPhysicalShapeLayer(Clip::hardEdge, true);
+  TestPhysicalShapeLayer(Clip::antiAlias, true);
+  TestPhysicalShapeLayer(Clip::antiAliasWithSaveLayer, false);
+}
+
+TEST(Layer, ShaderMaskSaveLayer) {
+  ShaderMaskLayer layer = ShaderMaskLayer(
+      SkShaders::Empty(), SkRect::MakeWH(5, 5), SkBlendMode::kSrcOver);
+  layer.Add(ReadbackLayer::Make(true, false));
+  EXPECT_FALSE(layer.tree_reads_surface());
+}
+
```

Comment:
Maybe also add a test for `BackdropFilterLayer` (the main reason why we're having this optimization)?

Reply:
True. I was focusing on making sure the known SaveLayer cases provided the expected protection...